library/ssl_tls.c

```diff
@@ -6618,7 +6589,9 @@ int mbedtls_ssl_read( mbedtls_ssl_context *ssl, unsigned char *buf, size_t len )
                 /* With DTLS, drop the packet (probably from last handshake) */
 #if defined(MBEDTLS_SSL_PROTO_DTLS)
                 if( ssl->conf->transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM )
-                    return( MBEDTLS_ERR_SSL_WANT_READ );
+                {
+                    continue;
+                }
```

Comment:
@RonEld The previous loop-driven example code should be ok both for the previous and the modified code base; for example, in `ssl_client2.c` there are two places where `mbedtls_ssl_read` is called, and they are both tightly wrapped in a loop repeating as long as `WANT_READ` or `WANT_WRITE` is returned. If one does it like this, blurring the two meanings of `WANT_READ` goes unnoticed, because in either case it's ok to immediately try `mbedtls_ssl_read` again. With the modified code, `mbedtls_ssl_read` will return `WANT_READ` fewer times, and when it does, it means that the the user can idle on the underlying transport to signal pending data, i.e. to employ event-driven IO - something that was not possible, or at least very error-prone, before.

So the example applications should stay fine the way they are, but it would be good to add sample code showing how to correctly use event-driven IO. I've made some changes in https://github.com/ARMmbed/mbedtls/pull/942 in that direction.